docs: transform and filter documentation (#2653)
kyma-project/telemetry-manager: docs/user/filter-and-process/ottl-transform-and-filter/README.md

# Transform and Filter with OTTL

To apply custom transformations and filters to your telemetry data, use the OpenTelemetry Transformation Language (OTTL). You can modify, enrich, and filter logs, metrics, and traces before they're sent to your backend.

## Overview

You define OTTL rules in the spec of your LogPipeline, MetricPipeline, and TracePipeline resources.

With the transform section, you can standardize and enrich your data: Add, rename, or modify fields to conform to a standard schema, add valuable context, or redact sensitive information. For details, see [Transform with OTTL](ottl-transform.md).

With the filter section, you can reduce data volume and cost: Drop unwanted logs, metrics, or traces to lower your ingestion costs and reduce noise in your observability backend. For details, see [Filter with OTTL](ottl-filter.md).

> When you define rules, the Telemetry module configures processors in the underlying OpenTelemetry Collector to run your statements.

Reviewer: ```suggestion
When you define rules, the Telemetry module configures processors in the underlying OpenTelemetry Collector to apply your rules.
```

transform processor has conditions and statements, while filter processor has only conditions. So, if we want this sentence to be generic for both features, then let's avoid saying `run your statements`

2. **OTTL Transformation**: Next, the pipeline applies the transformation rules from the transform section. These rules modify the data that passed the input filters. You can add, rename, or delete fields.
3. **OTTL Filtering**: Finally, the pipeline evaluates the transformed data against the filter rules from the filter section. Any data matching a filter condition is dropped.

This sequence means that your OTTL rules only operate on data that has already passed the initial input filters. Furthermore, your OTTL filter conditions must use the final, transformed state of your data, not its original state. For example, if you rename a field in a transformation rule, your OTTL filter must use the new name.

By default, if an OTTL statement encounters an error, the processor logs the error and continues to process the next piece of data. This ignore mode prevents a single malformed data point from stopping the entire pipeline.

## Limitations

### Always Use the Full Context Path

> You must specify the full path for every field, starting from the top-level context (resource, span, log, and so on). Short-hand references are not supported.

Reviewer: ```suggestion
You must specify the full path for every field. Short-hand references are not supported.
```

What do you mean by starting from the top-level context? For example, If I want to access the datapoint attributes, I will directly use `datapoint.attributes[""]` and datapoint is not the the top-level context in the metric. So, I will remove this part to avoid confusion

Author: I agree

> - Correct: 'resource.attributes["k8s.namespace.name"] == "default"', `log.attributes["level"]`, `datapoint.value`, `span.name`

Reviewer: ```suggestion
- Correct: `resource.attributes["k8s.namespace.name"] == "default"`, `log.attributes["level"]`, `datapoint.value`, `span.name`
```

> - Correct: 'resource.attributes["k8s.namespace.name"] == "default"', `log.attributes["level"]`, `datapoint.value`, `span.name`

Author: I will double-check

> - Incorrect: 'attributes["k8s.namespace.name"] == "default"', `attributes["level"]`, `value`, `name`

Reviewer: ```suggestion
- Incorrect: `attributes["k8s.namespace.name"] == "default"`, `attributes["level"]`, `value`, `name`
```

> [!TIP]
> For details on the underlying implementation details of context inference, see [OTel Transform Processor Context Inference](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/transformprocessor/README.md#context-inference).

### Trace Filtering Applies to Entire Spans

When you write a filter for a trace pipeline, the condition applies to the parent span. If the condition matches, the entire span is dropped, including all of its events.

> You cannot use filters to remove individual events within a span (the `spanevent` context).

Reviewer: ```suggestion
You cannot use filters to remove individual events within a span (the [spanevent context](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/contexts/ottlspanevent/README.md)).
```

2. **OTTL Transformation**: Next, the pipeline applies the transformation rules from the transform section. These rules modify the data that passed the input filters. You can add, rename, or delete attributes.
3. **OTTL Filtering**: Finally, the pipeline evaluates the transformed data against the filter rules from the filter section. Any data matching a filter condition is dropped.

This sequence means that your OTTL rules only operate on data that has already passed the initial input filters. Furthermore, your OTTL filter conditions must use the final, transformed state of your data, not its original state. For example, if you rename an attribute in a transformation rule, your OTTL filter must use the new name.

By default, if an OTTL statement encounters an error, the processor logs the error and continues to process the next piece of data. This ignore mode prevents a single malformed data point from stopping the entire pipeline.

> ## Predefined Contexts

Reviewer 2: excellent title, summarizes it nicely
